File: quick_sort_step.java
import java.io.*;
import java.util.*;

public class quick_sort_step {

    static List<String> steps = new ArrayList<>();

    public static void main(String[] args) throws Exception {
        if (args.length != 3) {
            System.out.println("Usage: java quick_sort_step <csv_file> <startRow> <endRow>");
            return;
        }

        String csvFile = args[0];
        int startRow = Integer.parseInt(args[1]);
        int endRow = Integer.parseInt(args[2]);

        List<Data> data = readDataset(csvFile, startRow, endRow);
        steps.add(formatList(data)); // Add initial state

        quickSort(data, 0, data.size() - 1);

        String outputFile = String.format("quick_sort_step_%d_%d.txt", startRow, endRow);
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(outputFile))) {
            for (String step : steps) {
                writer.write(step + "\n");
            }
        }

        System.out.println("Steps saved to " + outputFile);
    }

    static class Data {
        int number;
        String text;

        Data(int number, String text) {
            this.number = number;
            this.text = text;
        }

        @Override
        public String toString() {
            return number + "/" + text;
        }
    }

    static List<Data> readDataset(String filename, int startRow, int endRow) throws Exception {
        List<Data> result = new ArrayList<>();
        BufferedReader reader = new BufferedReader(new FileReader(filename));
        String line;
        int lineNum = 0;

        while ((line = reader.readLine()) != null) {
            lineNum++;
            if (lineNum >= startRow && lineNum <= endRow) {
                String[] parts = line.split(",");
                result.add(new Data(Integer.parseInt(parts[0]), parts[1]));
            }
        }

        reader.close();
        return result;
    }

    static void quickSort(List<Data> arr, int low, int high) {
        if (low < high) {
            int pi = partition(arr, low, high);
            steps.add("pi=" + pi + " " + formatList(arr));
            quickSort(arr, low, pi - 1);
            quickSort(arr, pi + 1, high);
        }
    }

    static int partition(List<Data> arr, int low, int high) {
        int pivot = arr.get(high).number;
        int i = low - 1;

        for (int j = low; j < high; j++) {
            if (arr.get(j).number < pivot) {
                i++;
                Collections.swap(arr, i, j);
            }
        }

        Collections.swap(arr, i + 1, high);
        return i + 1;
    }

    static String formatList(List<Data> arr) {
        List<String> formatted = new ArrayList<>();
        for (Data d : arr) {
            formatted.add(d.toString());
        }
        return formatted.toString();
    }
}

File: quick_sort_step.py
import csv
import sys

# Read a specific range of rows from the dataset
def read_rows(filename, start_row, end_row):
    rows = []
    with open(filename, 'r') as file:
        reader = csv.reader(file)
        for i, row in enumerate(reader):
            if start_row - 1 <= i <= end_row - 1:
                rows.append((int(row[0]), row[1]))
    return rows

# Recursive quick sort with step logging
def quick_sort_with_steps(arr, low, high, steps):
    if low < high:
        pi = partition(arr, low, high, steps)
        steps.append(f"pi={pi} " + str(format_data(arr)))
        quick_sort_with_steps(arr, low, pi - 1, steps)
        quick_sort_with_steps(arr, pi + 1, high, steps)

# Partition function (last element as pivot)
def partition(arr, low, high, steps):
    pivot = arr[high][0]
    i = low - 1
    for j in range(low, high):
        if arr[j][0] < pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]
    arr[i + 1], arr[high] = arr[high], arr[i + 1]
    return i + 1

# Format array for logging
def format_data(arr):
    return ["{}/{}".format(x[0], x[1]) for x in arr]

def main():
    if len(sys.argv) != 4:
        print("Usage: python quick_sort_step.py <csv_file> <start_row> <end_row>")
        return

    filename = sys.argv[1]
    start_row = int(sys.argv[2])
    end_row = int(sys.argv[3])

    data = read_rows(filename, start_row, end_row)

    steps = [str(format_data(data))]  # initial array
    quick_sort_with_steps(data, 0, len(data) - 1, steps)

    output_file = f"quick_sort_step_{start_row}_{end_row}.txt"
    with open(output_file, 'w') as out:
        for line in steps:
            out.write(line + '\n')

    print(f"Steps saved to {output_file}")

if __name__ == "__main__":
    main()
